Enable lnd pod logging

macgyver13 opened the pull request:
This update provides a new strategy for adding containers to a lnd pod via `extraContainers`. A reference implementation is integrated into `test/data/ln` and `test/data/logging`.

- Add lnd-exporter metric capture to test/data/ln (activated by **_metricsExport: true_**)
- Extend lnd pod template to support extraContainers
- Removes broken reference to circuitBreaker

Commit:
Add lnd exporter test data
Extend lnd pod template with extraContainers

Commit:
set "lnd" as default k8s container to support kubectl exec commands

Commit:
correctly define default annotation

willcl-ark commented:
<!--e57a25ab6845829454e8d69fc972939a-->

The following sections might be updated with supplementary metadata relevant to reviewers and maintainers.

<!--174a7506f384e20aa4161008e828411d-->
### Conflicts
No conflicts as of last run.

macgyver13 commented:
> The following sections might be updated with supplementary metadata relevant to reviewers and maintainers.
> 
> ### Conflicts
> Reviewers, this pull request conflicts with the following ones:
> 
> * [#688](https://github.com/bitcoin-dev-project/warnet/pull/688) (Circuit breaker plugin by Camillarhi)
> 
> If you consider this pull request important, please also help to review the conflicting pull requests. Ideally, start with the one that should be merged first.

#691 should proceed before #688 IMO

pinheadmz commented on docs/logging_monitoring.md:
```diff
@@ -116,6 +116,17 @@ outbounds 0.0
+lnd-exporter
+
+[lnd-exporter configuration reference](https://github.com/macgyver13/lnd-exporter/tree/main?tab=readme-ov-file#configuration)
+lnd-exporter assumes same macarroon referenced in ln_framework (can be overridden by env variable)
```
nit: macaroon

macgyver13 replied:
Is there a method for running all tests locally? I did not see anything so I created a test_runner that reads the .github/workflow/test.yml file and executes each test, is that of value to other developers?

ln_basic_test currently adds lnd-exporter container but does not enable scraping.  I think that type of tests would align better with logging_test than services_test, added lnd_block_height metric test to logging_test.  Please let me know if you prefer a different solution.

pinheadmz replied:
> Is there a method for running all tests locally?

No, mainly because on github actions everything can run in parallel and some of these tests take a long time, but a test runner would be a nice contribution if you are able to share!

macgyver13 replied:
I'll add another PR with local test runner and supporting documentation to keep this PR from sprawling.

pinheadmz reviewed (commented):
This is looking good so far. Try to make as few changes as possible to `test/data/ln` since there are tests that expect that network to have specific properties. Ideally you'd just be able to add exporters to the existing LN nodes in that network and then add a test (either in `ln_test.py` or '`ln_basic_test.py`) similar to the fork observer check in `services_test.py`

I'll try to allow-list you so CI runs automatically when you push. Be sure to run the tests locally as well!

Commit:
- Add grafana metric validation test to logging_test
- Reduced complexity of ln network.yaml config to align with ln_basic_test

Commit:
build lnd-exporter docker on linux - version 0.1.2

pinheadmz commented on test/data/logging/network.yaml:
```diff
@@ -11,5 +11,25 @@ nodes:
+          imagePullPolicy: IfNotPresent
+          env:
+            - name: ADMIN_MACAROON_HEX
+              value: 0201036c6e6402f801030a1062beabbf2a614b112128afa0c0b4fdd61201301a160a0761646472657373120472656164120577726974651a130a04696e666f120472656164120577726974651a170a08696e766f69636573120472656164120577726974651a210a086d616361726f6f6e120867656e6572617465120472656164120577726974651a160a076d657373616765120472656164120577726974651a170a086f6666636861696e120472656164120577726974651a160a07 ...
```
for a follow up PR maybe we can figure out how to make this value (and the tls.cert) a global value for all helm charts in the system

macgyver13 replied:
lnd-exporter assumes this value by default so it can technically be removed everywhere in warnet where lnd-exporter is referenced - just added for completeness.

pinheadmz replied:
Oh hm. Probably not the best idea to maintain a constant value in two separate repos, in case it ever changes in warnet ...?

macgyver13 replied:
Agree that is messy so lets go with your global helm chart approach, I know that is already implemented for chain (regtest)

m3dwards replied:
I don't really understand what the global helm approach would be but if it can be deleted then let's do that?

pinheadmz reviewed (commented):
Nice!!

![Screenshot 2025-03-12 at 10 56 56 AM](https://github.com/user-attachments/assets/c3bf61f7-f5eb-48a1-aad5-1eb6a2c4a90d)

pinheadmz commented on test/data/ln/node-defaults.yaml:
```diff
@@ -1,8 +1,40 @@
+          value: 0201036c6e6402f801030a1062beabbf2a614b112128afa0c0b4fdd61201301a160a0761646472657373120472656164120577726974651a130a04696e666f120472656164120577726974651a170a08696e766f69636573120472656164120577726974651a210a086d616361726f6f6e120867656e6572617465120472656164120577726974651a160a076d657373616765120472656164120577726974651a170a086f6666636861696e120472656164120577726974651a160a076f6e ...
+        - name: METRICS
+          value: >
+            lnd_balance_channels=parse("/v1/balance/channels","balance") 
+            lnd_local_balance_channels=parse("/v1/balance/channels","local_balance.sat") 
+            lnd_remote_balance_channels=parse("/v1/balance/channels","remote_balance.sat") 
+            lnd_block_height=parse("/v1/getinfo","block_height") 
+            lnd_peers=parse("/v1/getinfo","num_peers")
```
This is really cool but I don't see a test for it. I think that's ok although I would like to at least have a grafana dashboard I can easily load to see these data. Could be a follow up PR

pinheadmz reviewed (approved):
ACK aa4d3b20873fb48bd7a3325f0fc74c74540312aa
<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK aa4d3b20873fb48bd7a3325f0fc74c74540312aa
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmfRoZwACgkQ5+KYS2KJ
yTqnxg/+KN6pfvIT6qTB57z745+ndDBiljs6GeG2DDI8TuiskFnRwcZqsZUGsN9I
OnFBNx4NbPdvLKTQAJHLm/OpGVvn51Nj28EeRbv85hOP9CYg1UVS3dShkHahf3I3
2IHSdNnEQDjOsMsPIV11e6VY2AXr57SkhhdFy+XWq5XK5XSFvOdsoguNeg8SWjYn
yr7JC5Pdt8q4XBWZvvThLYGw/iRGdF6qnmEwYuvjD95Ik2d7XtCple8rcRMaCRkF
JrqlHCZCsaxfrViY5slpO92sxvgqGD/A0eGh5OJUZ0i+WXMhlNHUqA9f15tZL3Bs
c1UXplm7gQ40T+7SyBtYLf/xqpIfUwzW4MNFh3xf+B+HSjn+WqltQL4QY0pz4qyV
TaMCJ8pd1cwbgqIVgSgFAPPQR3YzkncdNfbm3IHRw74rlqkf+WdZIgp4X/wgA5Sv
hmpTYPZ0FHCrNilJMlGruXlPD1Fr2MXLDFyq06w8W8yG1fb2YVHaGDWil44ftsRs
Z4P7mWt+bafBgeNsisnaAlsV+htGUuLjlPyxD3GgMZBysmuQsTlBpJ27tzA6oC4f
qK5qTCSOVrkQGOIWHRhP5FmdqUOT0nT/tDSEEALpyq5R/oCP/CJJacG1aeLPbiA8
CH//oVf2G0tbbSRiuStwzk5CYF7XtFfBxlZcJBrHU5k8K0G+tR8=
=81wU
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on openpgp.org](https://keys.openpgp.org/vks/v1/by-fingerprint/E61773CD6E01040E2F1BD78CE7E2984B6289C93A)
</details>

m3dwards commented:
utACK aa4d3b20873fb48bd7a3325f0fc74c74540312aa

Commit:
update to support single macaroon reference
-shared volume for containers in same pod
-lookup configmap based on label for separate pods

Commit:
sanity check re-run pipeline - unable to reproduce error on local computer.
pipeline artifacts did not provide any useful information

pinheadmz commented on test/data/ln/node-defaults.yaml:
```diff
@@ -1,8 +1,40 @@
+    color=#000000
+  config: |
+    bitcoin.timelockdelta=33
+  metricsExport: false
```
why wouldn't this be `true` since you have metrics specified below?

pinheadmz replied:
I guess this doesn't matter for this PR anyway

pinheadmz commented:
ACK this is great thanks!

I added #696 locally and  will probably merge them together. Then I viewed the lnd dashboard after making a few changes to the `test/data/ln` network so there'd be something to look at:

```diff
diff --git a/resources/plugins/simln/charts/simln/templates/pod.yaml b/resources/plugins/simln/charts/simln/templates/pod.yaml
index 69790c9e..bda0344f 100644
--- a/resources/plugins/simln/charts/simln/templates/pod.yaml
+++ b/resources/plugins/simln/charts/simln/templates/pod.yaml
@@ -36,7 +36,7 @@ spec:
       args:
         - >
           cd /working;
-          sim-cli
+          sim-cli --capacity-multiplier=30
       volumeMounts:
         - name: {{ .Values.workingVolume.name }}
           mountPath: {{ .Values.workingVolume.mountPath }}
diff --git a/test/data/ln/network.yaml b/test/data/ln/network.yaml
index 5d30686d..bbf07636 100644
--- a/test/data/ln/network.yaml
+++ b/test/data/ln/network.yaml
@@ -19,8 +19,8 @@ nodes:
             block: 300
             index: 1
           target: tank-0004-ln
-          capacity: 100000
-          push_amt: 50000
+          capacity: 10000000
+          push_amt: 500000
   - name: tank-0004
     addnode:
       - tank-0000
@@ -30,8 +30,16 @@ nodes:
             block: 300
             index: 2
           target: tank-0005-ln
-          capacity: 50000
-          push_amt: 25000
+          capacity: 5000000
+          push_amt: 250000
   - name: tank-0005
     addnode:
-      - tank-0000
\ No newline at end of file
+      - tank-0000
+
+plugins:
+  postDeploy:
+    simln:
+      entrypoint: "../../../resources/plugins/simln"
+
+caddy:
+  enabled: true
\ No newline at end of file
diff --git a/test/data/ln/node-defaults.yaml b/test/data/ln/node-defaults.yaml
index 62e05199..d82ef96f 100644
--- a/test/data/ln/node-defaults.yaml
+++ b/test/data/ln/node-defaults.yaml
@@ -5,8 +5,8 @@ image:
   repository: bitcoindevproject/bitcoin
   pullPolicy: IfNotPresent
   tag: "27.0"
-collectLogs: false
-metricsExport: false
+collectLogs: true
+metricsExport: true
 
 #LN configs
 ln:
@@ -16,7 +16,7 @@ lnd:
     color=#000000
   config: |
     bitcoin.timelockdelta=33
-  metricsExport: false
+  metricsExport: true
   prometheusMetricsPort: 9332
   extraContainers:
     - name: lnd-exporter
```
